design-proposals/vm-pool.md

```diff
@@ -0,0 +1,194 @@
+* The VMConfig has a different status section from VM
+* A VMConfig’s name does not correlate in any way to a VM generated from the VMConfig
+* In a VMConfig, all naming resulting in newly created objects are treated as generated names in practice. For example, the name of a DataVolumeTemplate in a VMConfig.Spec.DataVolumeTemplates list will be treated as a generated name when VMs are spawned using the config. This ensures all VMs derived from a VMConfig are unique.
+
+**Similarities**
+* VMConfig and VM object’s spec structure are identical
+
+## VirtualMachinePool (VMPool) API
+
+The VMPool api represents all the tunings necessary for managing a pool of stateful VMs. This API is very simple and leverages the VMConfig object for most of the heavy lifting pertaining to what the VM instances within the pool actually look like.
+
+The VMPools spec contains the following tunings and values
+
+* **VirtualMachineRef** - (Required) A structure that points to a VMConfig object
+* **Replicas** - (Required) An integer representing the desired number of VM replicas
+* **MaxUnavailable**  - (Optional) (Defaults to 25%) Integer or string pointer, that when set represents either a percentage or number of VMs in a pool that can be unavailable (ready condition false) at a time during automated update.
+* **NameGeneration** - (Optional) Specifies how objects within a pool have their names generated
+	* **VMPrefix (Defaults to VMPool.Name)** - String representing the prefix to use for when naming VMs.
+	* **AppendPostfixToSecretReferences: (default false)** - Boolean that indicates if VM’s unique postfix should be appended to references to Secrets in the VMI’s Volumes list. This is useful when needing to pre-generate unique secrets for VMs within a pool.
+	* **AppendPostfixToConfigMapReferences (default false)** - Boolean that indicates if VM’s unique postfix should be appended to ConfigMap references in the VMI’s Volumes list. This is useful when needing to pre-generate unique secrets for VMs within a pool.
+* **DisruptionStrategy** - (Optional) Specifies how the VMPool controller manages scaling in and updating VMs within a VMPool
+	* **Unmanaged** - No automation during scale-in or updates. The VM is never touched after creation. Users manually delete and update individual VMs in a pool. Persistent state preservation is up to the user removing the VMs
+	* **Opportunistic** - Opportunistic scale-in and update of VMs which are in a halted state.
+		* **ScaleInStatePreservation** - (Optional) specifies if and how to preserve state of VMs selected for scale-in.
+			* **Disabled** - (Default) all state for VMs selected for scale-in will be deleted
+			* **Offline** - PVCs for VMs selected for scale-in will be preserved and reused on scale-out (decreases provisioning time during scale out)
```

Comment:
Would that mean name generation has to be consistent so the generated names for connected resources (storage) match a scaled-in VM when it happens?
From what I see above name generation would be incremental so it should be fine, just want to make sure.